### ISSUE TITLE
Error "Invalid type registration: :def_e" while usign mutant w/ rails application

### ISSUE THREAD

**Issue by kleber-gueriero** [mbj/mutant#1007]
While trying to run the rspec(not even calling the mutant command), like `$ rspec ./spec`, it raises an error:
```
An error occurred while loading ./spec/....
Failure/Error: require File.expand_path('../config/environment', __dir__)

Mutant::Registry::RegistryError:
  Invalid type registration: :def_e
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/registry.rb:25:in `register'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator.rb:26:in `block in handle'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator.rb:25:in `each'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator.rb:25:in `handle'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator/node/generic.rb:78:in `<class:Generic>'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator/node/generic.rb:8:in `<class:Node>'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator/node/generic.rb:5:in `<class:Mutator>'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator/node/generic.rb:4:in `<module:Mutant>'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant/mutator/node/generic.rb:3:in `<top (required)>'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
# /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
# /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
# /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# /usr/local/bundle/gems/mutant-0.9.7/lib/mutant.rb:75:in `<top (required)>'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
# /usr/local/bundle/gems/bootsnap-1.4.7/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
# ./config/application.rb:18:in `<top (required)>'
# ./config/environment.rb:2:in `require_relative'
# ./config/environment.rb:2:in `<top (required)>'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./spec/....rb:4:in `<top (required)>'
```

Ruby version: ruby 2.7.1p83(but also tested w/ 2.6.3, before it)
gem 'rails', '~> 5.2.3'
Using the open-source version of mutant

**Comment by mbj** [comment]
@kleber-gueriero I'm surprised that mutant gets loaded in this scenario. Must be "bootsnap beeing helpful".

Can you tell me the exact mutant version you are using? Must be one of the latest.

Overall there are 2 possible issues:

* Bootsnap tries to load mutant when its not supposed to, this could be under your control in your Gemfile etc.
* Mutant registers a mutator for the `def_e` node to its engine, but the parser version you are on does not provie it. This could be a fault in mutant not nailing the parser version too well.

As you are using the open-source version of mutant your code must also be opensource / in a public repo, can you point me to the repo, would help me a lot to track it down.

**Comment by mbj** [comment]
LOL yes you are on latest, the path tell me. So its likely that I need to be more conservative when checking the node registrations against the parser types. And allow mismatches if I support more than one minor release of parser.

**Comment by mbj** [comment]
@kleber-gueriero Okay so can you:

* post your `Gemfie` section around mutant?
* post your parser version?

**Comment by kleber-gueriero** [comment]
@mbj The Gemfile.lock part around mutant, w/ parser version:

```
mutant (0.9.7)
      abstract_type (~> 0.0.7)
      adamantium (~> 0.2.0)
      anima (~> 0.3.1)
      ast (~> 2.2)
      concord (~> 0.1.5)
      diff-lcs (= 1.3)
      equalizer (~> 0.0.9)
      ice_nine (~> 0.11.1)
      memoizable (~> 0.4.2)
      mprelude (~> 0.1.0)
      parser (~> 2.7.1)
      procto (~> 0.0.2)
      unparser (~> 0.4.6)
      variable (~> 0.0.1)
    mutant-rspec (0.9.5)
      mutant (~> 0.9.5)
      rspec-core (>= 3.8.0, < 4.0.0)
```

**Comment by mbj** [comment]
@kleber-gueriero can you check with master? The problem mutant fails to load itself in your bundle should be gone.

I cannot really help with the bootsnap issue, where bootsnap too eagerly loads mutant while you just intend to run rspec. For that one I really need to be pointed to the git repository.

**Comment by mbj** [comment]
@kleber-gueriero can you try the just released version 0.9.8 and tell me if mutant now loads correctly in your environment?
